storage: consistency check failed on cyan

https://sentry.io/cockroach-labs/cockroachdb/issues/666877375/

```
*log.safeError: replica_consistency.go:142: consistency check failed with %d inconsistent replicas | int
  File "github.com/cockroachdb/cockroach/pkg/storage/replica_consistency.go", line 151, in CheckConsistency
  File "github.com/cockroachdb/cockroach/pkg/storage/consistency_queue.go", line 114, in process
  File "github.com/cockroachdb/cockroach/pkg/storage/queue.go", line 744, in processReplica
  File "github.com/cockroachdb/cockroach/pkg/storage/queue.go", line 627, in 2
  File "github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go", line 324, in func1

replica_consistency.go:142: consistency check failed with %d inconsistent replicas | int
```

Heh, looks like the improved  roachtest reproduced this right away. Analysis incoming. First of all, I'll note that this failed relatively late in the merge process:

```
[   1] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (34s)
[   2] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (1m34s)
[   3] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (0.5%|2m34s)
[   4] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (2.5%|3m34s)
[   5] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (4.9%|4m34s)
[   6] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (7.5%|5m34s)
[   7] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (10.1%|6m34s)
[   8] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (12.7%|7m34s)
[   9] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (15.2%|8m34s)
[  10] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (17.7%|9m34s)
[  11] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (20.2%|10m34s)
[  12] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (22.4%|11m34s)
[  13] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (24.7%|12m34s)
[  14] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (27.0%|13m34s)
[  15] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (29.3%|14m34s)
[  16] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (31.4%|15m34s)
[  17] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (33.5%|16m34s)
[  18] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (35.8%|17m34s)
[  19] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (38.1%|18m34s)
[  20] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (40.4%|19m34s)
[  21] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (42.8%|20m34s)
[  22] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (45.3%|21m34s)
[  23] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (47.5%|22m34s)
[  24] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (49.8%|23m34s)
[  25] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (52.1%|24m34s)
[  26] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (54.5%|25m34s)
[  27] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (56.9%|26m34s)
[  28] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (59.2%|27m34s)
[  29] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (61.4%|28m34s)
[  30] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (63.8%|29m34s)
[  31] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (65.9%|30m34s)
[  32] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (68.2%|31m34s)
[  33] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (70.6%|32m34s)
[  34] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (72.9%|33m34s)
[  35] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (75.2%|34m34s)
[  36] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (77.3%|35m34s)
[  37] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (79.4%|36m34s)
[  38] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (81.8%|37m34s)
[  39] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (84.0%|38m34s)
[  40] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (86.0%|39m34s)
[  41] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (87.8%|40m34s)
[  42] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (89.2%|41m34s)
[  43] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (90.4%|42m34s)
[  44] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (91.6%|43m34s)
[  45] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (92.6%|44m34s)
[  46] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (93.6%|45m34s)
[  47] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (94.3%|46m34s)
[  48] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (95.2%|47m34s)
[  49] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (96.1%|48m34s)
[  50] clearrange: waiting for ~21896 merges to complete (and for at least 45m0s to pass) (97.0%|49m34s)
MONITOR: 8: dead
> /home/tschottdorf/go/bin/roachprod get tobias-wipe logs artifacts/clearrange/logs
tobias-wipe: getting (scp) logs artifacts/clearrange/logs

   1: done
   2: done
   3: done
   4: done
   5: done
   6: done
   7: done
   8: done
   9: done
  10: done
not destroying cluster to allow debugging
--- FAIL: clearrange [unstable] (3048.53s)
        test.go:498,cluster.go:1227,clearrange.go:166: unexpected node event: 8: dead
PASS (1 unstable FAIL)
```

This is perhaps a pattern. In cyan, the check also failed extremely late, when the range encompassed almost "everything".

--- a/pkg/storage/replica.go
+++ b/pkg/storage/replica.go
@@ -1717,6 +1717,11 @@ func (r *Replica) setDescWithoutProcessUpdate(desc *roachpb.RangeDescriptor) {
 		log.Fatalf(ctx, "cannot replace initialized descriptor with uninitialized one: %+v -> %+v",
 			r.mu.state.Desc, desc)
 	}
+	if r.mu.state.Desc != nil && !r.mu.state.Desc.StartKey.Equal(desc.StartKey) {
+		ctx := r.AnnotateCtx(context.TODO())
+		log.Fatalf(ctx, "attempted to change replica's start key from %s to %s",
+			r.mu.state.Desc.StartKey, desc.StartKey)
+	}
 
 	newMaxID := maxReplicaID(desc)
 	if newMaxID > r.mu.lastReplicaAdded {

--- a/pkg/storage/replica_raftstorage.go
+++ b/pkg/storage/replica_raftstorage.go
@@ -945,6 +945,9 @@ func (r *Replica) applySnapshot(
 	// Update the range and store stats.
 	r.store.metrics.subtractMVCCStats(*r.mu.state.Stats)
 	r.store.metrics.addMVCCStats(*s.Stats)
+	// TODO(benesch): the next line updates r.mu.state.Desc, but that's supposed
+	// to be handled by the call to setDescWithoutProcessUpdate below. This is not
+	// a correctness issue right now, but it's liable to become one.
 	r.mu.state = s
 	r.assertStateLocked(ctx, r.store.Engine())
 	r.mu.Unlock()

--- a/pkg/storage/store_snapshot.go
+++ b/pkg/storage/store_snapshot.go
@@ -394,7 +394,7 @@ func (s *Store) canApplySnapshot(
 func (s *Store) canApplySnapshotLocked(
 	ctx context.Context, snapHeader *SnapshotRequest_Header,
 ) (*ReplicaPlaceholder, error) {
-	desc := snapHeader.State.Desc
+	desc := *snapHeader.State.Desc
 	if v, ok := s.mu.replicas.Load(int64(desc.RangeID)); ok && (*Replica)(v).IsInitialized() {
 		// We have an initialized replica. Preemptive snapshots can be applied with
 		// no further checks if they do not widen the existing replica. Raft
@@ -420,7 +420,7 @@ func (s *Store) canApplySnapshotLocked(
 
 	// TODO(benesch): consider discovering and GC'ing *all* overlapping ranges,
 	// not just the first one that getOverlappingKeyRangeLocked happens to return.
-	if exRange := s.getOverlappingKeyRangeLocked(desc); exRange != nil {
+	if exRange := s.getOverlappingKeyRangeLocked(&desc); exRange != nil {
 		// We have a conflicting range, so we must block the snapshot.
 		// When such a conflict exists, it will be resolved by one range
 		// either being split or garbage collected.
@@ -454,7 +454,7 @@ func (s *Store) canApplySnapshotLocked(
 	}
 
 	placeholder := &ReplicaPlaceholder{
-		rangeDesc: *desc,
+		rangeDesc: desc,
 	}
 	return placeholder, nil
 }

